Add chromatic visual tests

tmeasday opened the pull request:
Issue: We didn't have visual regression tests for our stories

Note: this is built on https://github.com/storybooks/storybook/pull/2504

## What I did

Added [chromatic](https://blog.hichroma.com/introducing-chromatic-ui-testing-for-react-c5cc01a79aaa) to our CI

## How to test

Make a UI change, check out the build at https://chromaticqa.com

<strike>**NOTE**: this is not currently accessible to members of the `storybooks` org, just sorting out org permissions.</strike>

Anyone in the `storybooks` org on GH should be able to access the project here: https://www.chromaticqa.com/builds?appId=5a375b97f4b14f0020b0cda3

tmeasday committed:
Added chromatic for main storybook

tmeasday committed:
Run chromatic tests from CI

tmeasday commented:
@ndelangen do you have the permission to grant access from the `storybooks` GH org?

tmeasday committed:
Need to add the job to the build

Hypnosphi commented:
Note that there's an ongoing work on adding visual testing functionality to storyshots: #2413

tmeasday commented:
@Hypnosphi 

> Note that there's an ongoing work on adding visual testing functionality to storyshots

Yes! I think this is great! I think however for a larger project like this one Chromatic can offer a little more on top of that, namely parallelized / cloud screenshots and an review flow UI. Interested to hear your thoughts.

igor-dv commented on package.json:
```diff
@@ -38,7 +38,9 @@
-    "repo-dirty-check": "node ./scripts/repo-dirty-check"
+    "repo-dirty-check": "node ./scripts/repo-dirty-check",
+    "storybook": "start-storybook -p 6006",
+    "chromatic": "chromatic test --storybook-addon --app-code sm3026v658i"
```
If it's not a public key, probably better hide it with the CI parameter

tmeasday replied:
Yeah, see the note on the commit ;)

igor-dv replied:
Hah, 😄

Hypnosphi commented:
Oh, that's actually pretty cool. I think there's a value in adding this for all the `cra-kitchen-sink` stories

tmeasday commented:
> I think there's a value in adding this for all the cra-kitchen-sink stories

Yes, I'll get this going once we sort out #2504

tmeasday committed:
Temp: start storybook directly

Different storybook app layers install the `start-storybook` command on top of each other.

We won't leave the storybook here, though.

tmeasday committed:
Move some things around a little

ndelangen commented:
Please rebase, so i can better judge this PR, looks all right.

tmeasday commented:
@ndelangen done.

Hypnosphi commented on .circleci/config.yml:
```diff
@@ -199,6 +199,26 @@ jobs:
+      - run:
+          name: "Visually test storybook"
+          command: |
+            cd examples/official-storybook; yarn chromatic
```
Please add a root-level shortcut script for that

Hypnosphi replied:
Maybe it should be even made an option in `scripts/test.js`

ndelangen replied:
Is this something we want to be able to run locally?

Hypnosphi replied:
I believe so, at least to update the refs

ndelangen replied:
I was under the impression chromatic provided a review system to decide if a change is OK or NOT_OK ?

ndelangen replied:
If the tests are run by each dev, how does chromatic overcome the rendering differences per system?

Hypnosphi replied:
AFAIU they actually run in cloud even when launched locally

Hypnosphi replied:
> chromatic provided a review system to decide if a change is OK or NOT_OK

That's cool that they do that! And I feel convenient doing that even before submitting a PR

tmeasday replied:
Yeah, you can run it locally and it does take the snapshots in the cloud still.

The only hiccup with this is the `CHROMATIC_APP_CODE` environment variable -- it will need to be set for the test to run locally. Any thoughts on how to make that easier for devs (you can find it in the settings page at https://www.chromaticqa.com/settings?appId=5a375b97f4b14f0020b0cda3)

tmeasday replied:
@Hypnosphi -- not sure how to best add it to the `scripts/test.js` command; it seems the way that is written a command either needs to be a jest project or a `npm test` command for some subdirectory. I'm not sure we want to make it `npm test` for `examples/official-storybook` though.

Should I change the script to support a full custom command? Perhaps in a separate PR?

ndelangen committed:
Merge branch 'release/3.3' into tmeasday/chromatic

ndelangen committed:
Merge branch 'release/3.3' into tmeasday/chromatic

ndelangen committed:
Merge branch 'master' into tmeasday/chromatic

# Conflicts:
#	yarn.lock

ndelangen committed:
UPDATE lockfile

tmeasday committed:
Merge remote-tracking branch 'origin/master' into tmeasday/chromatic

tmeasday committed:
Added top-level chromatic command

tmeasday committed:
Don't fail the test if there are changes

Commit:
Merge branch 'master' into tmeasday/chromatic

tmeasday commented:
This is ready to merge when the 3.3 release process is complete.

Hypnosphi commented:
How does it affect release process?

Hypnosphi committed:
Merge branch 'master' into tmeasday/chromatic

tmeasday commented:
It shouldn't, but @shilman asked me to hold off merging into `master` until the release is out...

Hypnosphi committed:
Merge branch 'master' into tmeasday/chromatic

ndelangen committed:
Merge branch 'master' into tmeasday/chromatic

ndelangen commented:
Merge when ready, if you ask me

Hypnosphi committed:
Merge branch 'master' into tmeasday/chromatic
